alphastats/tl/differential_expression_analysis.py

```diff
@@ -0,0 +1,345 @@
+from abc import ABC, abstractmethod
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy
+from statsmodels.stats.multitest import multipletests
+
+from alphastats.dataset.keys import Cols, ConstantsClass
+from alphastats.dataset.preprocessing import PreprocessingStateKeys
+from alphastats.statistics.statistic_utils import calculate_foldchange
+
+
+class DeaColumns(ConstantsClass):
+    PVALUE = "p-value"
+    QVALUE = "q-value"
+    LOG2FC = "log2(fold change)"
+    SIGNIFICANT = "significant"
+
+
+class DeaParameters(ConstantsClass):
+    METADATA = "metadata"
+
+    GROUP1 = "group1"
+    GROUP2 = "group2"
+    GROUPING_COLUMN = "grouping_column"
+
+    TEST_FUN = "test_fun"
+    FDR_METHOD = "fdr_method"
+
+
+class DifferentialExpressionAnalysis(ABC):
+    """This class implements the basic methods required for differential expression analysis.
+
+    The purpose of this class is to provide a common interface for differential expression analysis. It should be subclassed for specific methods, such as t-tests or ANOVA. The class provides methods for input validation, output validation, and running the analysis. It also provides a method for getting the significance of the results based on a q-value cutoff.
+
+    attributes:
+    - input_data (pd.DataFrame): The input data for the analysis.
+    - result (pd.DataFrame): The result of the analysis.
+
+    abstract methods:
+    - _allowed_parameters: that returns a list of allowed parameters for the analysis, static method
+    - _extend_validation: extends the validation of parameters for the specific method
+    - _run_statistical_test: wrapper runs the statistical test from kwargs
+    - _statistical_test_fun: that runs the statistical test, static method. The wrapper and the actual method are separated to allow for easier testing and to ensure that all parameters are defined with types and defaults.
```

Comment:
I thought it would be nice to get an overview of the abstract mathods that need to be implemented by children. I'll move it down if this is the optionsal part.

Reply:
+1, I like it being visible in the class definition. Although keeping it up to date would mean double bookkeeping

Reply:
the double bookkeeping is my concern here.. and in this case, it's quite easy to find out what needs to be implemented: the `ABC` decorator takes care of that ;-)